File: min_stack.js
function minStack() {
  this.stack = [];
  this.minimumValue = null;
}

minStack.prototype.push = function(val) {
  //check if value is less than min
  //set min 
  if (this.minimumValue === null || val < this.minimumValue) {
    this.setMin(val);
  }
  this.stack.push(val);
};

minStack.prototype.pop = function(val) {
  this.stack.pop();
};

minStack.prototype.setMin = function(val) {
  this.minimumValue = val;
};

//return minimum value in O(1)
minStack.prototype.min = function() {
  console.log(this.minimumValue);
  return this.minimumValue;
};

File: min_stack.rb
# Instant lookup of minimum value in stack 
# require 'pry'
class MinStack
  def initialize 
    @stack = []
    @min = nil
  end 

  def pop 
    @stack.pop 
  end 

  def push(val)
    if (min.nil? || val < min)
      @min=val
    end 
    @stack.push(val)
  end 

  def min=(val)
    @min = val 
  end 

  def min 
    @min
  end 
end 

# x = MinStack.new
# p x 
# binding.pry 